### PR TITLE
Redesign QueryContext class

### PR DESCRIPTION
**Pull request by paul-rogers**
Query context revision which builds on the work started in [PR #13022](https://github.com/apache/druid/pull/13022). 

### Release Notes

Most of the work in this PR is transparent to users. This PR includes [Issue #13120](https://github.com/apache/druid/issues/13120)  which introduces a user-visible change. We introduce two new configuration keys that refine the query context security model controlled by `druid.auth.authorizeQueryContextParams`. When that value is set to `true` then two other configuration options become available:

* `druid.auth.unsecuredContextKeys`: The set of query context keys that *do not* require a security check. Use this for the "white-list" of key to allow. All other keys go through the existing context key security checks.
* `druid.auth.securedContextKeys`: The set of query context keys that *do* require a security check. Use this when you want to allow all but a specific set of keys: only these keys go through the existing context key security checks.

Both are set using JSON list format:

```text
druid.auth.securedContextKeys=["secretKey1", "secretKey2"]
```

You generally set one or the other values. If both are set, `unsecuredContextKeys` acts as exceptions to `securedContextKeys`.

In addition, Druid defines two query context keys which always bypass checks because Druid uses them internally:

* `sqlQueryId`
* `sqlStringifyArrays`

#### Backward Compatibility

When upgrading Druid, if query context security is disabled (`druid.auth.authorizeQueryContextParams=false`, the default) then you will see no change.

If query context security is enabled (`druid.auth.authorizeQueryContextParams=true`), then the two keys listed above will always be allowed, regardless of any security rules that may have been set. You can remove these two keys from your rules if you have added them.

The two new config values (`unsecuredContextKeys`, and `securedContextKeys`) will not be set in your configuration (or in the default configurations) and will thus behavior after an upgrade will not change. You can, however, shift context key security rules you have set for all users into the new `unsecuredContextKeys` config property.

### Background

Recent work uncovered two issues with the way Druid handles query contexts:

* #13049 fixes a race condition with JDBC updates
* #13022 fixes type-unsafe value conversions

During the investigation of the above it became clear that the semantics around the query context are a bit unclear and could be improved.

In particular, the current `QueryContext` class promises more than it can deliver. It groups context values into three groups: default, system values and user values. Splitting up the values was done to allow authorizing just the user-provided keys. However, this didn't quite work as intended for several reasons. First, most code does not recognize this format, and instead "merges" all three maps before making additional changes. Second, the only use of the user-provided values is to authorize them, something that can be done more simply via other means. Third, the concept of "user-set" keys is incorrect: keys set *may* come from the user, or may come from an application or the Router.

As a result, the `QueryContext` class creates confusion rather than tidying things up as was intended. The tidying up requires that we pull in [Issue #13120](https://github.com/apache/druid/issues/13120) to resolve a context key security ambiguity.

### Overview of Revisions

This PR achieves the desired goals in a simpler way, while unifying the many ways that the code currently works with the query context.

* The original three-map `QueryContext` is replaced by a "facade" on top of a single map.
* The original `QueryContext` was created to allow authorizing user context keys. This PR solves that issue by making a copy of the keys separate from the context, allowing the SQL engine to modify the context without creating conflicts with the security feature.
* Uses of `QueryContext` to fetch values are replaced with a reference the new `QueryContext` class as a facade.
* Modification to the context in the planning layer occurs directly on the context map; it is no longer done via the old `QueryContext` class since there is no real value in creating a layer on top of the map.
* #13022 adds a number of type-safe method to the `Query` interface. However, doing so clutters up that interface, and we still need the `QueryContexts` methods to get at specific context values. See the discussion in #13022 for details. This PR, by contrast, puts all context access methods on the revised `QueryContext` facade, making for cleaner code.
* All type-unsafe context value accesses were replaced in #13022. They are replaced again here by the new `QueryContext` type-safe methods.
* The old type-unsafe methods are marked deprecated. No code in Druid-proper references them, though we must assume that extensions may continue to use them.
* All specific methods in `QueryContexts` which take a `Query` or context map parameter are replaced by methods on the revised `QueryContext` class.
* `QueryContexts` methods which rewrote a query moved to the `Queries` class.
* Validation of a context value now reliably throws a `BadQueryContextException` with a clear explanation of the problem. Previously, we threw both `BadQueryContextException` and `ISE` for such errors, and the errors have variations in format and wording.
* `get` methods are standardized. `get<Type>()` without a default returns a boxed value, or `null` if the value is not set. This allows code to implement logic of the form "if the value is set, do something with the value, else do something else."
* `get(defaultValue)` methods return an unboxed value, since they will return the default if the value is unset.
* `DruidPlanner` is refactored a bit to better handle context key security.
* `AuthConfig` implements the enhanced context key security model.

### Specifics

The original `QueryContext` class was first removed: all references to that class were modified to refer to the context map instead.

A new class of the same name was created. The only creator of the new class is `Query.queryContext()`. The new class holds onto the context map associated with the query, and provides the many type-safe and value-specific methods. Since `QueryContext` is simply a facade, it is cheap to create an instance as needed. This lets us convert code of the form:

```java
long foo = QueryContexts.getAsLong(query, FOO_NAME, FOO_DEFAULT);
long bar = QueryContexts.getBar(query);
```

To

```java
long foo = query.queryContext().getLong(FOO_NAME, FOO_DEFAULT);
long bar = query.queryContext().getBar();
```

When a bit of code makes multiple references to context values (as above), we can also do:

```
QueryContext queryContext = query.queryContext();
long foo = queryContext.getLong(FOO_NAME, FOO_DEFAULT);
long bar = queryContext.getBar();
```

The following methods in `Query` are marked deprecated:

```java
interface Query... {
 @Deprecated
  <ContextType> ContextType getContextValue(String key);
  @Deprecated
  <ContextType> ContextType getContextValue(String key, ContextType defaultValue);
  @Deprecated
  default boolean getContextBoolean(String key, boolean defaultValue);
```

Although the above methods are deprecated, they are reimplemented as `default` methods. As a result, the methods of the same names are removed from subclasses since they are redundant.

The `Query.getQueryContext()` method is removed as a way of marking that the old `QueryContext` is no longer available. The old method didn't really even work: the different context types were lost when merging into a single map. A new `queryContext()` method returns the facade version. 

### Alternatives

The discussions in #13049 and #13022 suggested other alternatives:

* Work only with the context map and the `QueryContexts` methods as shown above. This works fine, and is one of the steps taken in #13049 that lead to this PR, but is a bit cumbersome.
* Add methods to the `Query` method as in an early draft of #13022. Doing so clutters the `Query` interface, and does not help in the SQL layer where we work with contexts before the native `Query` is created. Unless we really want to clutter the `Query` interface, we'd still need the value-specific methods in `QueryContexts`.
* Leave well enough alone: status quo. While the existing code works, as noted above, the semantics are a bit messy which creates confusion for readers. Tidying up the code has no benefit to the computer, but is helpful for us poor humans who have to understand the code.

### Conclusion

The result is a solution that:

* Solves the authorization issue that gave rise to the original `QueryContext`
* Solves the synchronization issue with `QueryContext` by avoiding multiple maps.
* Builds on the type safety solution of PR #13022 by shifting the methods to `QueryContext`.

<hr>

This PR has:
- [X] been self-reviewed.
- [X] has been validated using existing tests.
- [X] has modified unit tests as needed to reflect changed methods, exceptions and semantics.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by FrankChen021** [comment]
I agree with your proposal on the new API form as

```java
long foo = query.queryContext().getLong(FOO_NAME, FOO_DEFAULT);
long bar = query.queryContext().getBar();
```

It makes the APIs much simpler and cleaner.

**Comment by paul-rogers** [comment]
@FrankChen021, thanks for the encouraging comment! Since this PR builds on the work you recently did, can you perhaps give it a review? This one touches many files and so is at risk of conflicts, so we might want to get it in if things look OK. Thanks!

**Comment by FrankChen021** [comment]
> @FrankChen021, thanks for the encouraging comment! Since this PR builds on the work you recently did, can you perhaps give it a review? This one touches many files and so is at risk of conflicts, so we might want to get it in if things look OK. Thanks! 

Sure. I will review it on this weekends.

File: processing/src/main/java/org/apache/druid/query/Query.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/query/Query.java:122]: `getQueryContext` seems to be removed entirely. would it not break extensions?

File: processing/src/main/java/org/apache/druid/query/QueryContext.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:63]: if its a new function, can we call it `asMap`

File: server/src/main/java/org/apache/druid/server/QueryLifecycle.java
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/server/QueryLifecycle.java:107]: can we store the keys themselves instead of storing the map?

Comment (paul-rogers) [reply server/src/main/java/org/apache/druid/server/QueryLifecycle.java:107]: Great suggestion. Done.

File: processing/src/main/java/org/apache/druid/query/Query.java
Comment (paul-rogers) [reply processing/src/main/java/org/apache/druid/query/Query.java:122]: Great question. This is a relatively new method, so my thought is that it is safe to remove. It would be hard to retain as the shape of the returned `QueryContext` class itself is different. By removing this method (and replacing it with `context()`) we create a clear signal to early adopters that the class did change, and can't be used to modify the context any longer. Also, all the methods to get values, which extensions might have called, have also moved.

So, question: is this going to be a problem for extensions? Is there an expectation that old extensions, without recompiling and making a few simple changes) continue to work? If so, then this whole approach is called into question: we'd have to keep the old code and duplicate it in the new version, or just keep the old code and live with the mess. Thoughts?

File: processing/src/main/java/org/apache/druid/query/Query.java
Comment (paul-rogers) [reply processing/src/main/java/org/apache/druid/query/Query.java:142]: Good catch. Removed.

File: processing/src/main/java/org/apache/druid/query/Query.java
Comment (paul-rogers) [reply processing/src/main/java/org/apache/druid/query/Query.java:167]: Ah, this is a tricky one! So, a recent commit renamed this method. Since it is now deprecated, and presumably used only by old code, I put the name back to what it was before that commit.

**Review by paul-rogers — COMMENTED** [review]
@abhishekagarwal87, thanks for your review! Addressed your questions and comments. There is one backward-compatibility issue that one of your questions raised: please let me know what you think about the issue.

**Review by FrankChen021 — COMMENTED** [review]
Thank you @paul-rogers for the hard work. 

Reviewed most of the files(excluding test cases). Here're some of suggestions besides above comments.

1. A force-push is not recommended to an open PR since it breaks the review state kept by the GitHub. See https://github.com/apache/druid/blob/master/CONTRIBUTING.md
    > Avoid rebasing and force pushes after submitting a pull request, since these make it difficult for reviewers to see what you've changed in response to their reviews. The Druid committer that merges your change will rebase and squash it into a single commit before committing it to master.

2. Seems like your IDE adds extra line between `import javax` and `import java` directives, these changes are unneccessary

3. Regarding the `QueryContext`, my opinion is that caller side should always see this type instead of seeing its underlying storage type which is `Map<String, Object>`. We can see we have `getContext()`, `context()`, and even `queryContextMap()`, these are a little confusion. 

4. Separation of the `QueryContext` and `Map<String, Object>` also bring another problem. There're a lot of call of `context().getXXX`, each call of `context()` generates a new fascade object `QueryContext`, I think it's a little bit waste of memory and unnessary.

**Comment by paul-rogers** [comment]
@FrankChen021, thanks much for your detailed review! Very helpful. To answer your specific questions:

1. Sorry for the force-push. On many projects, the build is done on the PR itself. Any conflicts are resolved at the end of the review process. In Druid, each PR build is first rebased on master. Failure to rebase causes the build to fail and we get no feedback on the code. As a result, when there is a merge conflict, we are forced to rebase, which requires a force push. I wonder, is there some trick that others use to avoid this issue?
2. Sorry for the extra spaces in imports. That's just how Eclipse works. I'll make a note to manually remove such newlines in the future.
3. See below.
4. See below.

Your point 3 suggests that callers always see `QueryContext` and point 4 worries about the cost of creating `QueryContext` per call. Looking more carefully at the prior implementation, there is a path to make this work. That fix appears in the latest commit. The result is we eliminate the copies that were your concern in point 4.

Point 3 also suggests that callers always see `QueryContext` instead of the map. I agree that is the goal. The one caveat is that, for backward compatibility, the original map-based `getContext()` must exist and be a JSON property so that the JSON form of a query is unchanged.

Thank you for pointing out the potential bug with updating a query context in place in a query. The general rule, after this change, should be:

* The query context is created, as a map, before the query is constructed.
* Once the query is constructed, the context is accessed via `QueryContext` and is immutable. (Again, the one exception is that the map is presented to Jackson for JSON serialization.)
* If the context must change (such as while processing a native query), we must make a copy of the context then make a copy of the entire query with the new context. This ensures that the query is immutable and avoids the kinds of race conditions which first brought my attention to this area.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/QueryHandler.java
Comment (paul-rogers) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/QueryHandler.java:539]: You've identified another nasty inconsistency.

Our general rule for context values seems to be that they should be in JSON-compatible format, since we use Jackson to deserialize. So, if this context value appears in the context as a JSON number, it must follow the JSON rules and will be converted to `Integer`, `Long`, `Float`, etc. depending on the format of the number. The usual `QueryContexts.getAsLong()` method seems to convert from whatever Jackson created to get an actual `Long`. This handles `x` where `x` is a number.

We also handle `"x"` where we quote the number from above. JSON says that a string is a string: there is no JSON-defined conversion from strings to numbers. If we want to allow that, then Druid has to provide the logic to do so. Unfortunately, before your fix (and my doubling-down on your approach), it looked like every bit of code could make up its own parsing rules.

We can perhaps standardize and say that, if we want to parse `{"foo": "x"}` as a number, then we want the same result as if the user had actually written `{"foo": x}`, where 'x` is a valid JSON number. This means that, when the value is a string, what we actually want to do to parse is:

```java
Long longValue = objectMapper.readValue(context.getString("x"), Long.class);
```

We probably don't actually want to do the above: it is costly and requires a mapper. But, we want our string-to-long conversion to produce the same result as the above. A new commit added tests to probe the odd cases to see how closely the context functions follow JSON and the dimension utilities.

As it turns out, both the dimension utilities and Jackson does handle a conversion from "20.0" to a Long, but the context `getAsLong()` does not. To fix this, I added the `convertObjectToLong()` code to `getAsLong()` so that the context methods also handle this case. That fix should address the concern you raised.

File: server/src/main/java/org/apache/druid/client/DirectDruidClient.java
Comment (paul-rogers) [reply server/src/main/java/org/apache/druid/client/DirectDruidClient.java:168]: This is a tricky one, hence the comment. The call to `queryContext.getLong(QUERY_FAIL_TIME)` looks up the value, and converts the value to `Long`. If the value is unset, we get a `null` `Long`. When we convert `(long) null` we get a null pointer exception.

In this case, that's OK because `QUERY_FAIL_TIME` is set by Druid itself, so we know it is set correctly. This is the wrong way to get user-provided values: we have to deal with a missing value, or provide a default.

The comment simply means that we thought about the issues, and if a developer sees an NPE, then check what changed in the code because the value should be set.

**Comment by FrankChen021** [comment]
> Sorry for the force-push. On many projects, the build is done on the PR itself. Any conflicts are resolved at the end of the review process. In Druid, each PR build is first rebased on master. Failure to rebase causes the build to fail and we get no feedback on the code. As a result, when there is a merge conflict, we are forced to rebase, which requires a force push. I wonder, is there some trick that others use to avoid this issue?

To resolve merge conflict, another way is to merge the latest master back to our own development branch.

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java
Comment (FrankChen021) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:376]:
```suggestion
   * @param value   if {@code null}, remove the key. Otherwise, insert or replace
```

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java
Comment (FrankChen021) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:386]: Any special reason to use `TreeMap` instead of `HashMap`?

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java
Comment (FrankChen021) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:381]: don't see the `context` is used inside this function, do I miss sth?

Comment (paul-rogers) [reply processing/src/main/java/org/apache/druid/query/QueryContexts.java:381]: No, I did. Thanks for catching.

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java
Comment (paul-rogers) [reply processing/src/main/java/org/apache/druid/query/QueryContexts.java:386]: Another good catch. Native queries are serialized. It is handy if that serialization is deterministic. A `HashMap` will return its key/value pairs in random order, but `TreeMap` returns them in sorted order.

Given that, the right place to create the `TreeMap` is in `QueryContext` itself. Here, I lazily copied what was done in another similar method. Since `QueryContext` now copies the context into a `TreeMap`, there is no need to use one here as well.

File: processing/src/main/java/org/apache/druid/query/Query.java
Comment (FrankChen021) [inline processing/src/main/java/org/apache/druid/query/Query.java:144]: There's still a reference in `SinkQuerySegmentWalker` to this deprecated interface

File: processing/src/main/java/org/apache/druid/query/QueryContext.java
Comment (FrankChen021) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:63]:
```suggestion
    this.context = context == null ? Collections.emptyMap() : Collections.unmodifiableMap(new TreeMap<>(context));
```

File: sql/src/main/java/org/apache/druid/sql/AbstractStatement.java
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/AbstractStatement.java:67]: So this is mutable?

File: sql/src/main/java/org/apache/druid/sql/SqlQueryPlus.java
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/SqlQueryPlus.java:52]: And this is immutable?

File: sql/src/main/java/org/apache/druid/sql/avatica/DruidConnection.java
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/avatica/DruidConnection.java:99]: Can we move the copy into the ctor of `DruidJdbcStatement` so that no one forgets to make a copy?

File: sql/src/main/java/org/apache/druid/sql/avatica/DruidConnection.java
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/avatica/DruidConnection.java:129]: No need to make a copy because the ctor of `SqlQueryPlus` always make a copy.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/sql/MSQModeTest.java
Comment (FrankChen021) [inline extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/sql/MSQModeTest.java:37]:
```suggestion
    Map<String, Object> originalQueryContext = new HashMap();
```

Assigned with `Collections.emptyMap` causes the UT failed.

File: sql/src/main/java/org/apache/druid/sql/AbstractStatement.java
Comment (paul-rogers) [reply sql/src/main/java/org/apache/druid/sql/AbstractStatement.java:67]: Yes. Revised the comment to make the semantics a bit clearer. See if the revised text helps.

**Review by paul-rogers — COMMENTED** [review]
@FrankChen021, thanks again for our very thorough review. I'm impressed at the detailed items you identified: you somehow find all the tender bits in the code, which is wonderful.

I did a re-inspection of each of the areas you identified. I'm hoping that, this time, the story is consistent in each area.

**Review by FrankChen021 — APPROVED** [review]

File: processing/src/main/java/org/apache/druid/query/Query.java
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/query/Query.java:122]: I think it's fine. I am just going to tag the PR "Backward incompatible"

Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/query/Query.java:122]: Please add a note in your PR that can be used in release notes later.

**Comment by paul-rogers** [comment]
@FrankChen021, thanks for the approval!

So, it turns out that I had to pull the work for [Issue #13120](https://github.com/apache/druid/issues/13120) into this PR in order to get the security IT to pass. Basically, that IT does context key security checks using JDBC. JDBC inserts a key, which fails the security check. This worked previously when we separated system and user keys. (In fact, this may be the reason that the prior PR did the split.)

This latest PR adds two config values, defined in the "release notes" section in the revised description. To make this work:

* The `AuthConfig` is not the one-stop shop to prepare the list of context keys to use for authorization checks.
* Three new set operations are added to `CollectionUtils` because Java, in its infinite wisdom, doesn't provide them.
* Two context keys moved from the SQL package to `QueryContexts` so that they are visible to `AuthConfig`.
* These two keys are "out-of-the-box freebies" for context security: they are always allowed because they are set by Druid itself in either the Router (query ID) or JDBC path (stringify arrays.)
* The code in the planner to gather the resources is tidied up a bit.

If we're good with the proposed changes, perhaps we can wrap up this PR and get it merged.

**Comment by paul-rogers** [comment]
PR is updated to pull the work for [Issue #13120](https://github.com/apache/druid/issues/13120). That change is needed to get the security IT to pass. Basically, that IT does context key security checks using JDBC. JDBC inserts a key, which fails the security check. This worked previously when we separated system and user keys. (In fact, this may be the reason that the prior PR did the split.)

This latest PR adds two config values, defined in the "release notes" section in the revised description. To make this work:

* The `AuthConfig` is not the one-stop shop to prepare the list of context keys to use for authorization checks.
* Three new set operations are added to `CollectionUtils` because Java, in its infinite wisdom, doesn't provide them.
* Two context keys moved from the SQL package to `QueryContexts` so that they are visible to `AuthConfig`.
* These two keys are "out-of-the-box freebies" for context security: they are always allowed because they are set by Druid itself in either the Router (query ID) or JDBC path (stringify arrays.)
* The code in the planner to gather the resources is tidied up a bit.

We'll want to review this final change before we merge.

File: server/src/main/java/org/apache/druid/server/security/AuthConfig.java
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/server/security/AuthConfig.java:164]: nit - could we call this method `contextKeysToAuthorize` or `filterContextKeysToAuthorize`?

File: server/src/main/java/org/apache/druid/server/security/AuthConfig.java
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/server/security/AuthConfig.java:119]: should this be initialized as empty map like unsecuredContextKeys?

**Review by abhishekagarwal87 — COMMENTED** [review]
Thank you, Paul. I don't see any change in the docs. will you add that in another PR?

**Comment by abhishekagarwal87** [comment]
It will also make life easier for cluster admin to have a default list of unsecured context variables.

**Comment by abhishekagarwal87** [comment]
There are legit CI failures.

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by abhishekagarwal87** [comment]
Thank you @paul-rogers. We can make the changes I suggested in a follow-up PR.

**Pull request merged**
